### PR TITLE
Hide user's emails to protect their privacy

### PR DESCRIPTION
**Pull request by scott97**
Users are now identified with a numeric ID and have a display name that is for displaying publicly. Roles have been refactored out into its own package. An alternative middleware NoAuth is added for when running locally. Annotations and videostreams now store a user ID rather than email. API docs have been updated.

image

### CODE REVIEW

File: cmd/openfish/handlers/annotations.go [hunk 1]
```diff
@@ -233,10 +233,7 @@ func CreateAnnotation(ctx *fiber.Ctx) error {
 	}
 
 	// Get logged in user.
-	observer := "testuser"
-	if ctx.Locals("email") != nil {
-		observer = ctx.Locals("email").(string)
-	}
+	observer := ctx.Locals("id").(int64)
```
Comment (ao-david) [inline cmd/openfish/handlers/annotations.go:236]: I'm probably missing it, but is this set somewhere?

Comment (ao-david) [reply cmd/openfish/handlers/annotations.go:236]: Also, is it standard practice to not check the type assertion?

Comment (scott97) [reply cmd/openfish/handlers/annotations.go]: It is not set and so it won't work.  I have fixed that  now. I am now checking the type assertions.

**Review by trekhopton — CHANGES_REQUESTED** [review]

File: cmd/openfish/services/users.go [hunk 2]
```diff
@@ -38,31 +38,117 @@ import (
...
+
+// UserContents is the contents of a user in private contexts.
+type UserContents struct {
+	Email       string    `json:"email" example:"coral.fischer@example.com"`
+	DisplayName string    `json:"display_name" example:"Coral Fischer"`
+	Role        role.Role `json:"role" swaggertype:"string" example:"annotator"`
+}
+
+// PartialUserContents is for updating a user with a partial update (such as a PATCH request).
+type PartialUserContents struct {
```
[the remark this answers is not in the document]
Comment (scott97) [reply cmd/openfish/services/users.go:67]: Yeah basically

File: cmd/openfish/handlers/annotations.go [hunk 3]
```diff
@@ -233,22 +233,25 @@ func CreateAnnotation(ctx *fiber.Ctx) error {
 	}
 
 	// Get logged in user.
-	observer := "testuser"
-	if ctx.Locals("email") != nil {
-		observer = ctx.Locals("email").(string)
+	observer, ok := ctx.Locals("user").(*services.User)
+	if !ok {
+		return fmt.Errorf("unknown error")
```
[the remark this answers is not in the document]
Comment (scott97) [reply cmd/openfish/handlers/annotations.go]: Done :+1:

File: cmd/openfish/services/users.go [hunk 4]
```diff
@@ -38,31 +38,117 @@ import (
...
+func (u *UserContents) ToEntity() entities.User {
+	e := entities.User{
+		Email:       u.Email,
+		DisplayName: u.DisplayName,
+		Role:        u.Role,
+	}
+
+	return e
+}
+
```
[the remark this answers is not in the document]
Comment (scott97) [reply cmd/openfish/services/users.go:92]: Done

File: cmd/openfish/handlers/auth.go [hunk 5]
```diff
@@ -35,22 +35,77 @@ LICENSE
 package handlers
 
 import (
-	"github.com/ausocean/openfish/cmd/openfish/entities"
+	"fmt"
+
+	"github.com/ausocean/openfish/cmd/openfish/api"
+	"github.com/ausocean/openfish/cmd/openfish/services"
+	"github.com/ausocean/openfish/cmd/openfish/types/role"
 	"github.com/gofiber/fiber/v2"
```
Comment (trekhopton) [inline cmd/openfish/handlers/auth.go:43]: Order of imports should be standard library, then 3rd party, then ausocean packages.

File: cmd/openfish/handlers/user.go [hunk 6]
```diff
@@ -35,74 +35,67 @@ LICENSE
...
 import (
+	"fmt"
+	"strconv"
+
 	"github.com/ausocean/openfish/cmd/openfish/api"
-	"github.com/ausocean/openfish/cmd/openfish/entities"
 	"github.com/ausocean/openfish/cmd/openfish/services"
+	"github.com/ausocean/openfish/cmd/openfish/types/role"
 
 	"github.com/gofiber/fiber/v2"
```
Comment (trekhopton) [inline cmd/openfish/handlers/user.go:45]: Order imports, std, 3rd party, ausocean

File: cmd/openfish/middleware/authentication.go [hunk 7]
```diff
@@ -38,12 +38,24 @@ import (
 	"fmt"
 
 	"github.com/ausocean/openfish/cmd/openfish/api"
-	"github.com/ausocean/openfish/cmd/openfish/entities"
 	"github.com/ausocean/openfish/cmd/openfish/services"
+	"github.com/ausocean/openfish/cmd/openfish/types/role"
 	"github.com/gofiber/fiber/v2"
 	"google.golang.org/api/idtoken"
```
Comment (trekhopton) [inline cmd/openfish/middleware/authentication.go:44]: Order imports, std, 3rd party, ausocean, separated by a line.

File: cmd/openfish/middleware/authentication.go [hunk 8]
```diff
@@ -62,42 +74,29 @@ func ValidateJWT(aud string) func(*fiber.Ctx) error {
...
-			role = user.Role
-		}
-
-		// Add subject, email, and user role to ctx.Locals.
-		ctx.Locals("subject", subject)
 		ctx.Locals("email", email)
-		ctx.Locals("role", role)
+
+		// Fetch user from datastore if they exist.
+		user, _ := services.GetUserByEmail(email)
```
Comment (trekhopton) [inline cmd/openfish/middleware/authentication.go:82]: Why did we get rid of the error handling?

File: cmd/openfish/services/users.go [hunk 9]
```diff
@@ -39,31 +39,119 @@ import (
 
 	"github.com/ausocean/openfish/cmd/openfish/ds_client"
 	"github.com/ausocean/openfish/cmd/openfish/entities"
+
```
Comment (trekhopton) [inline cmd/openfish/services/users.go:42]: Remove extra line.

**Review by trekhopton — CHANGES_REQUESTED** [review]
Minor comments, checks still need to pass.

File: cmd/openfish/handlers/auth.go (hunk 5 above)
Comment (scott97) [reply cmd/openfish/handlers/auth.go:43]: When I run `go fmt ./cmd/openfish` it reorders them into this order, so I would prefer not to change that.

File: cmd/openfish/handlers/user.go (hunk 6 above)
Comment (scott97) [reply cmd/openfish/handlers/user.go:45]: When I run go fmt ./cmd/openfish it reorders them into this order, so I would prefer not to change that.

File: cmd/openfish/middleware/authentication.go (hunk 7 above)
Comment (scott97) [reply cmd/openfish/middleware/authentication.go:44]: When I run go fmt ./cmd/openfish it reorders them into this order, so I would prefer not to change that.

File: cmd/openfish/middleware/authentication.go (hunk 8 above)
Comment (scott97) [reply cmd/openfish/middleware/authentication.go]: fixed

**Review by trekhopton — APPROVED** [review]

**Pull request merged**